llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -5300,6 +5347,15 @@ bool CombinerHelper::matchUDivByConst(MachineInstr &MI) {
   if (MF.getFunction().hasMinSize())
     return false;
 
+  if (MI.getFlag(MachineInstr::MIFlag::IsExact)) {
+    return matchUnaryPredicate(
+        MRI, RHS, [](const Constant *C) { return C && !C->isNullValue(); });
+  }
+
+  auto *RHSDef = MRI.getVRegDef(RHS);
+  if (!isConstantOrConstantVector(*RHSDef, MRI))
+    return false;
+
   // Don't do this if the types are not going to be legal.
   if (LI) {
     if (!isLegalOrBeforeLegalizer({TargetOpcode::G_MUL, {DstTy, DstTy}}))
```

Comment:
Do we need to check that G_MUL and G_LSHR are legal before doing your new transform?

Reply:
Oh I guess we didn't check them for the equivalent sdiv transform so maybe not?

Reply:
We do not because exact division transform does not do an extending multiplication (umull for example).